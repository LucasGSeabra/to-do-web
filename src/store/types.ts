import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { RootState, AppDispatch } from './store';

export interface todo {
  _id: string
  done: boolean
  description: string
  createdAt: string
}

export interface todoState {
  description: string
  list: todo[]
}

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
