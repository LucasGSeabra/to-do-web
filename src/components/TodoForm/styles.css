.todoForm {
    margin-bottom: 10px !important;
}
