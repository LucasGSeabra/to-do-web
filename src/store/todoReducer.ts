import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
import { todoState, todo } from './types';

const URL = 'https://glacial-castle-07790.herokuapp.com/api/todos';

const initialState: todoState = {
  description: '',
  list: [],
};

export const fetchTodos = createAsyncThunk('fetchTodos', async (parameter: string) => {
  const search = parameter ? `&description__regex=/${parameter}/` : '';
  const response = await axios.get(`${URL}?sort=-createdAt${search}`);
  return response.data;
});

export const addTodos = createAsyncThunk('addTodos', async (parameter: string) => {
  const description = parameter;
  const response = await axios.post(URL, { description });
  return response.data;
});

export const changeStatus = createAsyncThunk('changeStatus', async (parameter: todo) => {
  const item = parameter;
  const response = await axios.put(`${URL}/${item._id}`, { ...item, done: !item.done });
  return response.data;
});

export const removeTodos = createAsyncThunk('removeTodos', async (parameter: string) => {
  const id = parameter;
  await axios.delete(`${URL}/${id}`);
  return id;
});

export const todoReducer = createSlice({
  name: 'todo',
  initialState,
  reducers: {
    descriptionChanged(state, action) {
      return { ...state, description: action.payload };
    },
    descriptionCleared(state) {
      return { ...state, description: '' };
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchTodos.fulfilled, (state, action) => ({ ...state, list: action.payload }))
      .addCase(addTodos.fulfilled, (state, action) => {
        const newList = state.list;
        newList.unshift(action.payload);
        state.description = '';
      })
      .addCase(changeStatus.fulfilled, (state, action) => {
        const { _id, done } = action.payload;
        const item = state.list.find((todoItem) => todoItem._id === _id);
        item!.done = done;
      })
      .addCase(removeTodos.fulfilled, (state, action) => {
        const id = action.payload;
        const newList = state.list.filter((todoItem) => todoItem._id !== id);
        return { ...state, list: newList };
      });
  },
});

export const { descriptionChanged, descriptionCleared } = todoReducer.actions;

export default todoReducer.reducer;
