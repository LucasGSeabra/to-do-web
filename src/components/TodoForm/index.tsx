import React from 'react';
import { Row, Col } from 'react-bootstrap';
import TodoButton from '../TodoButton';
import FormInput from '../FormInput';
import './styles.css';

function TodoForm() {
  return (
    <form className="todoForm">
      <Row>
        <Col xs={12} md={10} sm={9}>
          <FormInput />
        </Col>
        <Col xs={12} md={2} sm={3}>
          <TodoButton variant="primary" />
          <TodoButton variant="info" />
          <TodoButton variant="secondary" />
        </Col>
      </Row>
    </form>
  );
}

export default TodoForm;
