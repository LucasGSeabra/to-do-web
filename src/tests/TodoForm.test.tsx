/* eslint-disable no-undef */
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import TodoForm from '../components/TodoForm';
import store from './testReducer';

describe('TodoForm Component', () => {
  it('should render TodoForm correctly', () => {
    const component = renderer.create(
      <Provider store={store}>
        <TodoForm />
      </Provider>,
    ).toJSON();
    expect(component).toMatchSnapshot();
  });
});
