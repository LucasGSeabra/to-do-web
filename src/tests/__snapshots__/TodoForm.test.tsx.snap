// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoForm Component should render TodoForm correctly 1`] = `
<form
  className="todoForm"
>
  <div
    className="row"
  >
    <div
      className="col-md-10 col-sm-9 col-12"
    >
      <input
        className="form-control"
        id="description"
        onChange={[Function]}
        onKeyUp={[Function]}
        placeholder="Adicione uma tarefa"
        value="Teste"
      />
    </div>
    <div
      className="col-md-2 col-sm-3 col-12"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-plus fa-w-14 "
          data-icon="plus"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="btn btn-info"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-search fa-w-16 "
          data-icon="search"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times fa-w-11 "
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </div>
</form>
`;
